[Feat] Button 컴포넌트 퍼블리싱

bang-wol opened the pull request:
## 🚀 작업 내용

- Button 공통 컴포넌트 구현 

## 📝 참고 사항

- `varient`, `shape`, `size` 옵션으로 스타일을 지정할 수 있습니다. 
- `type`은 button으로 기본 설정 되어있습니다. 

## 🖼️ 스크린샷

<img width="783" alt="image" src="https://github.com/user-attachments/assets/2c6c5037-bea1-4b0b-9d4b-238c10872c84">


## 🚨 관련 이슈

- close #7 

## ✅ 체크리스트

- [x] Code Review 요청
- [x] Label 설정
- [x] PR 제목 규칙에 맞는지 확인

bang-wol committed:
Feat: Button 컴포넌트 구현

bang-wol committed:
Test: Button 컴포넌트 테스트 페이지

Young2un commented on src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,72 @@
+  variant = 'filled',
+  shape = 'rounded',
+  size = 'default',
+  disabled = false,
```
P5: 혹시 이거 안써주면 오류가 날까용?

bang-wol replied:
disabled 자체를 말씀하시는 걸까요 아니면.. false만 안 써주는 걸 말씀하시는 걸까요!?
disabled는 스타일 지정해주는데 필요해서 있어야 하는데, false로 값 설정은 안 해줘도 될 것 같네용!

Young2un commented on src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,72 @@
+export default function Button({
+  className,
+  children,
+  type,
```
P5 여기도 기본값으로 줘도 될거 같아요!
`type = 'button'`

bang-wol replied:
넵!! 기본 값으로 줘야 될 것 같네요! 기본 값이 button인 줄 알았는데 submit이네용..

Young2un commented on src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,72 @@
+  return (
+    <button
+      className={combinedClasses}
+      type={type === 'submit' ? 'submit' : 'button'}
```
P5 아마 린트가 꼭 타입 쓰라 해서 이렇게 하신거 같은데!
```suggestion
      type={type}
```
이렇게 하는게 조금 나을 거 같아요 버튼타입이 하나 더 있긴 해성,,
[버튼 타입 공식문서](https://developer.mozilla.org/ko/docs/Web/HTML/Element/button#type)
```
버튼의 행동 방식. 가능한 값은 다음과 같습니다.

submit: 버튼이 서버로 양식 데이터를 제출합니다. 지정하지 않은 경우 기본값이며, 유효하지 않은 값일 때도 사용합니다.
reset: <input type="reset">처럼, 모든 컨트롤을 초깃값으로 되돌립니다.
button: 기본 행동이 없으며 클릭했을 때 아무것도 하지 않습니다. 클라이언트측 스크립트와 연결할 수 있습니다.
```

bang-wol replied:
저두 처음에는 말씀해주신 것처럼 type={type} 이렇게 썼었는데 린트가 이것도 오류로 잡더라구요...

[Disallow usage of button elements without an explicit type attribute (react/button-has-type)](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/button-has-type.md
)

`var Hello = <button type={condition ? "button" : "submit"}>Hello</button>`
이것도 올바른 코드라고 하니.. 제가 쓴 코드 그대로 해도 될 것 같기도 합니닷...

Young2un reviewed (approved):
와아우,, 아주 깔꼼합니당!! 베리굿이에용!!!

bang-wol committed:
Fix: props 기본값 수정

sockki commented on src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,72 @@
+  shape?: 'rectangle' | 'rounded';
+  size?: 'small' | 'default' | 'large';
+}
+
```
P5: type 속성 전부다 옵셔널 프로퍼티를 주셨는데, 저거 3개 다 전달하지 않으면 어떻게 표시되나요? 테스트 페이지에 variant랑 shape는 다 전달 해주셔서 혹시 저 두개는 옵셔널 프로퍼티 빼도 괜찮을지 여쭈어 봅니다.

bang-wol replied:
제가 버튼 기본값으로  
```
 type = 'button',
 variant = 'filled',
 shape = 'rounded',
 size = 'default',
```
이렇게 다 설정해두어서 아무런 옵션을 주지 않아도 기본값으로 설정해둔 스타일이 적용됩니다!
기본값 지정해두고 옵셔널로 설정해두면 많은 옵션을 입력하지 않아도 편하게 사용할 수 있을 것 같아서 옵셔널로 해두었습니다!

sockki reviewed (approved):
고생하셨습니다. 구조 짜느라 머리 아프셨겠네요.

rl0425 commented on src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,72 @@
+  return (
+    <button
+      className={combinedClasses}
+      type={type === 'submit' ? 'submit' : 'button'}
```
type의 기본값이 'button' 인 것 같은데,  불필요한 조건문인 것 같습니다.
```typescript
type={type}
``` 
그대로 사용하셔도 될 것 같습니다.

bang-wol replied:
<img width="940" alt="스크린샷 2024-08-18 오후 8 58 46" src="https://github.com/user-attachments/assets/e73a4a51-1b40-4926-b175-a8fb5cc57ae7">

위에 영은님께서도 말씀해주셔서 답변했는데 `type={type}`으로 작성하면 이렇게 eslint 오류가 발생해서 조건문을 작성했습니다.

printjin-gmailcom commented:
고생하셨습니다.

bang-wol committed:
Chore: Merge branch 'develop' into 7-feat-button-컴포넌트-퍼블리싱

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

bang-wol committed:
Fix: lint 오류 해결

bang-wol committed:
Test: Button 컴포넌트 스토리북 작성

bang-wol committed:
Refactor: use client 지시문 추가

bang-wol committed:
Refactor: interface 사용을 type 사용으로 변경

bang-wol committed:
Style: config에 정의된 색상 이름으로 변경
